includes/acl/clip_reader/clip_reader.h

```diff
@@ -0,0 +1,450 @@
+#include "acl/clip_reader/clip_reader_error.h"
+#include "acl/compression/animation_clip.h"
+#include "acl/sjson/sjson_parser.h"
+#include <functional>
+
+namespace acl
+{
+	class ClipReader
+	{
+	public:
+		ClipReader(Allocator& allocator, const char* const sjson_input, int input_length)
+			: m_allocator(allocator),
+			  m_parser(sjson_input, input_length),
+			  m_error()
+		{
+		}
+
+		~ClipReader()
+		{
+			m_allocator.deallocate(m_bone_names);
+		}
+
+		bool read()
+		{
+			if (m_read_already)
+			{
+				return false;
+			}
+
+			m_read_already = true;
+
+			if (read_version() &&
+				read_clip_header() &&
+				read_and_allocate_bones() &&
+				create_clip() &&
+				read_tracks() &&
+				nothing_follows())
+			{
+				return true;
+			}
+
+			if (m_clip != nullptr)
+			{
+				m_clip->~AnimationClip();
+				m_allocator.deallocate(m_clip);
+				m_clip = nullptr;
+			}
+
+			if (m_skeleton != nullptr)
+			{
+				m_skeleton->~RigidSkeleton();
+				m_allocator.deallocate(m_skeleton);
+				m_skeleton = nullptr;
+			}
+
+			return false;
+		}
+
+		AnimationClip* get_clip()
+		{
+			return m_clip;
+		}
+
+		RigidSkeleton* get_skeleton()
+		{
+			return m_skeleton;
+		}
+
+		ClipReaderError get_error()
+		{
+			return m_error;
+		}
+
+	private:
+		// Temporary storage for input that must be translated to compressor data structures.
+		struct Header
+		{
+			const char* clip_name;
+			int clip_name_length;
+			int num_samples;
+			int sample_rate;
+			double error_threshold;
+			const char* reference_frame;
+			int reference_frame_length;
+		};
+
+		struct Bone
+		{
+			int index;
+			const char* name;
+			int name_length;
+			const char* parent;
+			int parent_length;
+			double vertex_distance;
+			double bind_rotation[4];
+			double bind_translation[3];
+			double bind_scale[3];
+		};
+
+		Allocator& m_allocator;
+		bool m_read_already{};
+		SJSONParser m_parser;
+		ClipReaderError m_error;
+
+		RigidSkeleton* m_skeleton{};
+		AnimationClip* m_clip{};
+
+		double m_version{};
+		Header m_header{};
+
+		int m_num_bones{};
+		char* m_bone_names{};
+
+		void set_error(int reason)
+		{
+			m_parser.get_position(m_error.line, m_error.column);
+			m_error.error = reason;
+		}
+
+		bool read_version()
+		{
+			if (!m_parser.read("version", m_version))
+			{
+				m_error = m_parser.get_error();
+				return false;
+			}
+
+			if (m_version != 1.0)
+			{
+				set_error(ClipReaderError::UnsupportedVersion);
+				return false;
+			}
+
+			return true;
+		}
+
+		bool read_clip_header()
+		{
+			if (!m_parser.object_begins("clip"))
+			{
+				goto error;
+			}
+
+			if (!m_parser.read("name", m_header.clip_name, m_header.clip_name_length) ||
+				!m_parser.read("num_samples", m_header.num_samples) ||
+				!m_parser.read("sample_rate", m_header.sample_rate) ||
+				!m_parser.read("error_threshold", m_header.error_threshold))
+			{
+				goto error;
+			}
+
+			m_parser.try_read("reference_frame", m_header.reference_frame, m_header.reference_frame_length);
```

Comment:
I actually removed this property from the format, I instead commented it, everything should be defined always in parent bone local space, just easier that way